safehttp/form.go

```diff
@@ -0,0 +1,252 @@
+// values. In case it does, it will try to convert the first value to an unsigned integer and
+// return it. If the first value is not an unsigned integer or there are no values
+// associated with paramKey, it will return the default value and set the Form
+// error field.
+func (f *Form) Uint(paramKey string, defaultValue uint64) uint64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramKey]
+	if !ok {
+		f.err = fmt.Errorf("no value found for key %q", paramKey)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseUint(vals[0], 10, 0)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// String checks whether key paramKey maps to any query or form parameter
+// values. In case it does, it will return the first value. If it does not, it
+// will return the default value and set the Form error field.
+func (f *Form) String(paramKey string, defaultValue string) string {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramKey]
+	if !ok {
+		f.err = fmt.Errorf("no value found for key %q", paramKey)
+		return defaultValue
+	}
+	return vals[0]
+}
+
+// Float64 checks whether key paramKey maps to any query or form parameter
+// values. In case it does, it will try toconvert the first value to a float and
+// return it. If the first value is not a float or there are no values
+// associated with paramKey, it will return the default value and set the Form
+// error field.
+func (f *Form) Float64(paramKey string, defaultValue float64) float64 {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramKey]
+	if !ok {
+		f.err = fmt.Errorf("no value found for key %q", paramKey)
+		return defaultValue
+	}
+	paramVal, err := strconv.ParseFloat(vals[0], 64)
+	if err != nil {
+		f.err = err
+		return defaultValue
+	}
+	return paramVal
+}
+
+// Bool checks whether key paramKey maps to any query  or form parameter
+// values. In case it does, it will try to convert the first value to a booleand
+// and
+// return it. If the first value is not a boolean or there are no values
+// associated with paramKey, it will return the default value and set the Form
+// error field.
+func (f *Form) Bool(paramKey string, defaultValue bool) bool {
+	if f.err != nil {
+		return defaultValue
+	}
+	vals, ok := f.values[paramKey]
+	if !ok {
+		f.err = fmt.Errorf("no value found for key %q", paramKey)
+		return defaultValue
+	}
+	switch vals[0] {
+	case "true":
+		return true
+	case "false":
+		return false
+	default:
+		f.err = fmt.Errorf("values of form parameter %q not a boolean", paramKey)
+	}
+	return false
+}
+
+// clearSlice is a helper functions that clears the slice slicePtr points to when an error occured
+func (f *Form) clearSlice(slicePtr interface{}) {
+	switch vs := slicePtr.(type) {
+	case *[]string:
+		*vs = nil
+	case *[]int:
+		*vs = nil
+	case *[]float64:
+		*vs = nil
+	case *[]uint64:
+		*vs = nil
+	case *[]bool:
```

Comment:
missed?

Reply:
wups, sorry - resolved in [ 43c86aa](url)